From dared1st@yahoo.com Tue Jul 11 23:38:44 2006
Message-ID: <20060712063841.18958.qmail@web81202.mail.mud.yahoo.com>
Date: Tue, 11 Jul 2006 23:38:41 -0700 (PDT)
From: Aleksey Gorelov <dared1st@yahoo.com>
Subject: USB: Properly unregister reboot notifier in case of failure in ehci hcd
Cc: gregkh@suse.de

If some problem occurs during ehci startup, for instance, request_irq fails, echi hcd driver
tries it best to cleanup, but fails to unregister reboot notifier, which in turn leads to crash on
reboot/poweroff. Below is the patch against current git to fix this.

I did not check if the same problem existed for uhci/ohci host drivers.

Signed-off-by: Aleks Gorelov <dared1st@yahoo.com>
Cc: David Brownell <david-b@pacbell.net>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

---
 drivers/usb/host/ehci-hcd.c |    7 +++----
 1 file changed, 3 insertions(+), 4 deletions(-)

--- gregkh-2.6.orig/drivers/usb/host/ehci-hcd.c
+++ gregkh-2.6/drivers/usb/host/ehci-hcd.c
@@ -483,9 +483,6 @@ static int ehci_init(struct usb_hcd *hcd
 	}
 	ehci->command = temp;
 
-	ehci->reboot_notifier.notifier_call = ehci_reboot;
-	register_reboot_notifier(&ehci->reboot_notifier);
-
 	return 0;
 }
 
@@ -499,7 +496,6 @@ static int ehci_run (struct usb_hcd *hcd
 
 	/* EHCI spec section 4.1 */
 	if ((retval = ehci_reset(ehci)) != 0) {
-		unregister_reboot_notifier(&ehci->reboot_notifier);
 		ehci_mem_cleanup(ehci);
 		return retval;
 	}
@@ -560,6 +556,9 @@ static int ehci_run (struct usb_hcd *hcd
 	 */
 	create_debug_files(ehci);
 
+	ehci->reboot_notifier.notifier_call = ehci_reboot;
+	register_reboot_notifier(&ehci->reboot_notifier);
+
 	return 0;
 }
 
